languages/ruby/exercises/concept/enumeration/.docs/introduction.md

```diff
@@ -0,0 +1,42 @@
+Enumeration is the act of stepping through a collection (Array, Hash, etc) and performing some action on each object.
+Enumeration is a key concept in Ruby and is used for sorting (`sort_by`), grouping (`group_by`), mapping (`map`), reducing (`reduce`), and much more. You'll most frequently see enumeration as the idiomatic way iterating through collections rather than using loops.
+
+A simple enumeration to print each word in an array would look like this:
+
+```ruby
+words = %w[the cat sat on the mat]
+words.each do |word| 
+  print word 
+end
+
+#=> the
+#=> cat
+#=> ...
+#=> mat
+```
+
+In this example, we have called the Array's `each` method and passed in a _block_, which takes one parameter (`word`) and prints it. Blocks are small groupings of statements that can be executed multiple times. They can be thought of as closures or anonymous functions. Blocks are defined using the `do...end` syntax (above), or the `{}` (below). The styles are interchangeable and differing opinions exist about when each should be used.
+
+We can also chain enumerable methods. For example, we can chain `.with_index` onto `each` to print out the index of an object as well as it's value:
+
+```ruby
+words = %w[the cat sat on the mat]
+words.each.with_index { |word, index| print "#{index}: #{word}" }
```

Comment:
```suggestion
words.map.with_index { |word, index| puts "#{index}: #{word}" }
```
Which will output:

```
0: the
1: cat
2: sat
3: on
4: the
5: mat
```
The colon is used here to avoid potential recognition of the numbers as "with a period".

But now "chain enumerable methods" is more accurate in what is happening here, we chain `map` (an `Enumerable` method) to `with_index` (another `Enumerable` method).